src/merge_events.hpp

```diff
@@ -0,0 +1,83 @@
+#pragma once
+
+#include <algorithm>
+#include <vector>
+
+#include <event_generator.hpp>
+#include <event_queue.hpp>
+#include <profiling/profiler.hpp>
+
+namespace arb {
+
+// merge_events generates a sorted list of postsynaptic events that are to be
+// delivered after the current epoch ends. It merges events from multiple
+// sources:
+//  lc : the list of currently enqueued events
+//  events : an unsorted list of events from the communicator
+//  generators : a set of event_generators
+//
+// The time intervales are illustrated below, along with the range of times
+// for events in each of lc, events and generators
+//  * t the start of the current epoch (not required to perform the merge).
+//  * t₀ the start of the next epoch
+//  * t₁ the end of the next epoch
+//
+//   t      t₀     t₁
+//   |------|------|
+//
+//   [----------------------] lc
+//          [---------------] events
+//          [------) generators
+//
+// The output list, stored in lf, will contain all the following:
+//  * all events in events list
+//  * events in lc with time >= t₀
+//  * events from each generator with time < t₁
+// All events in lc that are to be delivered before t₀ are discared, along with
```

Comment:
This comment appears to contradict itself. events on generators are discarded. But the next event on the generator is one that is discarded.

Reply:
No events from generators are discarded in practice: `merge_events` takes all events from the generators with delivery time < `t1`. It stops reading events from the generators when it finds the first event in a generator that is to be delivered on or after `t1`. That event will be taken on the next call to `merge_events`, in the following epoch.
So you are right, this comment is a bit confusing.